h2o-core/src/main/java/water/rapids/ast/prims/mungers/AstGroup.java

```diff
@@ -515,4 +589,91 @@ public String toString() {
       return Arrays.toString(_gs);
     }
   }
+
+  // --------------------------------------------------------------------------
+  // For each groupG and each aggregate function (median only), we separate and
+  // extract the column per groupG per aggregate function into a NewChunk column
+  // here.
+  private static class BuildGroup extends MRTask<BuildGroup> {
+    final int[] _gbCols;
+    private final AGG[] _aggs;   // Aggregate descriptions
+    private final int _medianCols;
+    IcedHashMap<G, String> _gss;
+    private G[] _grps;
+
+
+    BuildGroup(int[] gbCols, AGG[] aggs, IcedHashMap<G, String> gss, G[] grps, int medianCols) {
+      _gbCols = gbCols;
+      _aggs = aggs;
+      _gss = gss;
+      _grps = grps;
+      _medianCols = medianCols;
+    }
+
+    @Override
+    public void map(Chunk[] cs, NewChunk[] ncs) {
+      G gWork = new G(_gbCols.length, _aggs, _medianCols); // Working Group
+      G gOld;
+
+      for (int row = 0; row < cs[0]._len; row++) {  // for each
+        gWork.fill(row, cs, _gbCols);
+        gOld = _gss.getk(gWork);
+        for (int i = 0; i < gOld._isMedian.length; i++) { // Accumulate aggregate reductions
+          if (gOld._isMedian[i]) {  // median action required on column and group
+            double d1 = cs[gOld._medianCols[i]].atd(row);
+            if (!Double.isNaN(d1) || gOld._na[i] != NAHandling.RM)
+              ncs[gOld._newChunkCols[i]].addNum(d1);  // build up dataset for each group
+          }
+        }
+      }
+    }
+
+    // For the data column collected for each G and each aggregate function, make a frame out of the data
+    // newChunk column.  Sort the column and return median as the middle value of mean of two middle values.
+    Vec[] close() {
+      Futures fs = new Futures();
+
+      int cCount = 0;
+      Vec[] tempVgrps = new Vec[_medianCols];
+      for (G oneG : _grps) {
+        for (int index = 0; index < oneG._isMedian.length; index++) {
+          if (oneG._isMedian[index]) {  // median action is needed
+            // make a frame out of the NewChunk vector
+            tempVgrps[cCount++] = _appendables[oneG._newChunkCols[index]].close(_appendables[oneG._newChunkCols[index]].compute_rowLayout(), fs);
+          }
+        }
+      }
+      fs.blockForPending();
+      return tempVgrps;
+    }
+
+    public void calcMedian(Vec[] tempVgrps) {
+      int cCount = 0;
+      for (G oneG : _grps) {
+        for (int index = 0; index < oneG._isMedian.length; index++) {
+          if (oneG._isMedian[index]) {
+            Vec[] vgrps = new Vec[1];
+            vgrps[0] = tempVgrps[cCount++];
+            Frame myFrame = new Frame(Key.<Frame>make(), vgrps, true);
```

Comment:
Does this frame need a key? Is it ever installed in DKV?

Reply:
Yes, the sort function needs a frame with a key in order to work.  Since the frame is passed as part of the argument, I assumed it does not need to be installed in the DKV.  What do you think?